pkg/detector/ospkg/wolfi/wolfi.go

```diff
@@ -0,0 +1,129 @@
+	"k8s.io/utils/clock"
+
+	dbTypes "github.com/aquasecurity/trivy-db/pkg/types"
+	"github.com/aquasecurity/trivy-db/pkg/vulnsrc/wolfi"
+	ftypes "github.com/aquasecurity/trivy/pkg/fanal/types"
+	"github.com/aquasecurity/trivy/pkg/log"
+	"github.com/aquasecurity/trivy/pkg/scanner/utils"
+	"github.com/aquasecurity/trivy/pkg/types"
+)
+
+type options struct {
+	clock clock.Clock
+}
+
+type option func(*options)
+
+func WithClock(clock clock.Clock) option {
+	return func(opts *options) {
+		opts.clock = clock
+	}
+}
+
+// Scanner implements the Wolfi scanner
+type Scanner struct {
+	vs wolfi.VulnSrc
+	*options
+}
+
+// NewScanner is the factory method for Scanner
+func NewScanner(opts ...option) *Scanner {
+	o := &options{
+		clock: clock.RealClock{},
+	}
+
+	for _, opt := range opts {
+		opt(o)
+	}
+	return &Scanner{
+		vs:      wolfi.NewVulnSrc(),
+		options: o,
+	}
+}
+
+// Detect vulnerabilities in package using Wolfi scanner
+func (s *Scanner) Detect(_ string, _ *ftypes.Repository, pkgs []ftypes.Package) ([]types.DetectedVulnerability, error) {
+	log.Logger.Info("Detecting Wolfi vulnerabilities...")
+
+	log.Logger.Debugf("wolfi: the number of packages: %d", len(pkgs))
+
+	var vulns []types.DetectedVulnerability
+	for _, pkg := range pkgs {
+		srcName := pkg.SrcName
+		if srcName == "" {
+			srcName = pkg.Name
+		}
+		advisories, err := s.vs.Get("", srcName)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to get Wolfi advisories: %w", err)
+		}
+
+		installed := utils.FormatSrcVersion(pkg)
+		installedVersion, err := version.NewVersion(installed)
+		if err != nil {
+			log.Logger.Debugf("failed to parse Wolfi Linux installed package version: %s", err)
+			continue
+		}
+
+		for _, adv := range advisories {
+			if !s.isVulnerable(installedVersion, adv) {
+				continue
+			}
+			vulns = append(vulns, types.DetectedVulnerability{
+				VulnerabilityID:  adv.VulnerabilityID,
+				PkgID:            pkg.ID,
+				PkgName:          pkg.Name,
+				InstalledVersion: installed,
+				FixedVersion:     adv.FixedVersion,
+				Layer:            pkg.Layer,
+				Ref:              pkg.Ref,
+				Custom:           adv.Custom,
+				DataSource:       adv.DataSource,
+			})
+		}
+	}
+	return vulns, nil
+}
+
+func (s *Scanner) isVulnerable(installedVersion version.Version, adv dbTypes.Advisory) bool {
+	// This logic is for unfixed vulnerabilities, but Trivy DB doesn't have advisories for unfixed vulnerabilities for now
+	// because Alpine just provides potentially vulnerable packages. It will cause a lot of false positives.
+	// This is for Aqua commercial products.
+	if adv.AffectedVersion != "" {
+		// AffectedVersion means which version introduced this vulnerability.
+		affectedVersion, err := version.NewVersion(adv.AffectedVersion)
+		if err != nil {
+			log.Logger.Debugf("failed to parse Wolfi Linux affected package version: %s", err)
+			return false
+		}
+		if affectedVersion.GreaterThan(installedVersion) {
+			return false
+		}
+	}
```

Comment:
Does Wolfi provide advisories for unfixed vulnerabilities? If no, we can remove these lines.

Reply:
Okay, thanks! Removed in 40c4eb3f